src/sources/dnstap/dns_message.rs

```diff
@@ -0,0 +1,287 @@
+#[readonly::make]
+#[derive(Clone, Debug)]
+pub struct DnsQueryMessage {
+    pub response_code: u16,
+    pub response: Option<&'static str>,
+    pub header: QueryHeader,
+    pub question_section: Vec<QueryQuestion>,
+    pub answer_section: Vec<DnsRecord>,
+    pub authority_section: Vec<DnsRecord>,
+    pub additional_section: Vec<DnsRecord>,
+    pub opt_pserdo_section: Option<OptPseudoSection>,
+}
+
+impl DnsQueryMessage {
+    pub fn new(
+        response_code: u16,
+        response: Option<&'static str>,
+        header: QueryHeader,
+        question_section: Vec<QueryQuestion>,
+        answer_section: Vec<DnsRecord>,
+        authority_section: Vec<DnsRecord>,
+        additional_section: Vec<DnsRecord>,
+        opt_pserdo_section: Option<OptPseudoSection>,
+    ) -> Self {
+        Self {
+            response_code,
+            response,
+            header,
+            question_section,
+            answer_section,
+            authority_section,
+            additional_section,
+            opt_pserdo_section,
+        }
+    }
+}
+
+#[readonly::make]
+#[derive(Clone, Debug)]
+pub struct QueryHeader {
+    pub id: u16,
+    pub opcode: u8,
+    pub rcode: u8,
+    pub qr: u8,
+    pub aa: bool,
+    pub tc: bool,
+    pub rd: bool,
+    pub ra: bool,
+    pub ad: bool,
+    pub cd: bool,
+    pub question_count: u16,
+    pub answer_count: u16,
+    pub authority_count: u16,
+    pub additional_count: u16,
+}
+
+impl QueryHeader {
+    pub fn new(
```

Comment:
It's not for a particular reason, just following suit of the old school principle of `least privilege` in an extensive way 🙂 .

Reply:
But since you've raised concern about it (and the concern looks quite reasonable to me too), I'll drop all of these `new()` methods from structs declared in `dns_message.rs`, and simply declare all struct fields as `pub` instead.